### ISSUE TITLE
compose_ids() not returning for email reply

### ISSUE THREAD

**Issue by Osmosis311** [KartikTalwar/gmail.js#404]
I'm using the latest version of GmailJS as of one week ago.  For new compose windows, compose_ids works fine, however, for replies/forwards, it's returning an empty array.

Can you check? Thanks!

![gmail_compose](https://user-images.githubusercontent.com/360240/27043373-f27931b8-4f67-11e7-86f6-2b409d0001da.png)

**Comment by josteink** [comment]
Sorry about the slow response.  I'll try to make up for that by at least being somewhat extensive.

First of all: I can reproduce this issue.

This hasn't really been an issue for me in my projects, since I intercept the compose instance in the following event instead:

````javascript
g.observe.on("compose", (compose_ref, type) => {
    console.log("Compose opened:", compose_ref, type);
});
//true
//Compose opened: api.dom.compose {$el: jQuery.fn.init(1)} reply
````

Replies are typically "inline" compose, and not floating on top like the "normal" ones. That's why `gmail.get.compose_ids()` aren't able to pick them up. I guess that's a bug we should somehow be able to solve?

If I "pop up" an empty reply compose (making it visually the same as a new compose), it *almost* shows up in `gmail.get.compose_ids()`:

````javascript
g.get.compose_ids();
// [""]
````

If I fill something into the reply before  popping it up, it works:

````javascript
g.get.compose_ids();
// ["15cd3b9561f6db38"]
````

This is consistent with new compose windows too: Unless you fill something out, it doesn't get saved as a draft. Thus Gmail has not yet assigned an ID to it:

![image](https://user-images.githubusercontent.com/411338/27469852-a75af6ca-57f1-11e7-8716-6ff55083f518.png)

What's interesting here is that Gmail seems to have its own compose-numbering system:

![image](https://user-images.githubusercontent.com/411338/27469924-eb7e7f5c-57f1-11e7-83ac-e7ac0ffff5bc.png)

@KartikTalwar: Do you know anything about these IDs? Are there any reason we prefer the draft-ID value over the compose-ID value?

I'll be honest and say that for me, this issue is not a priority to solve, and I'm not going to put any direct effort into it. If you however would like to take a stab at it, I'll be more than willing to test and review any PRs you may have.

So basically I guess you have two options:

* Rewrite your code to instead depend on `compose`-events (these work fairly reliably).
* Send in a PR to also detect "inline" compose-windows.
* If needed change compose-ID system to use compose-ID value instead of draft-ID value.

Hopefully you'll find a solution which works for you.

**Comment by josteink** [comment]
To add to this. From the docs:

> gmail.get.compose_ids()
>
> Returns the latest/last email id of emails that have been saved as drafts (currently open)
>
> ["14469c73bba6ff53", "14469c7b8581ccfa"]

As such it makes sense we only pick up things saved as drafts. The following change is sufficient for me to pick up both "inline" and "floating" compose-windows:

````javascript
    api.get.compose_ids = function() {
        var ret = [];
        var dom = $(".M9 [name=draft]");
        for(var i = 0; i < dom.length; i++) {
            if(dom[i].value !== "undefined"){
                ret.push(dom[i].value);
            }
        }
        return ret;
    };
````

Basically just replace the `.AD`-selector with `.M9` instead.

@KartikTalwar: I see `M9` used a lot elsewhere. Any reason it's not used here?

**Comment by KartikTalwar** [comment]
@josteink no specific reason comes to mind for me. I believe @brentkelly implemented most of the compose functions

**Comment by josteink** [comment]
So unless anyone else have any objections, I'll just go ahead and make that change then?

**Comment by Osmosis311** [comment]
Looks good to me! Excited for this to be implemented!  Thanks!

**Comment by josteink** [comment]
I've [pushed the fix to master](https://github.com/KartikTalwar/gmail.js/commit/54ab61932d5522421b7535c10b6a303b97ac3a43) and bumped the version.

@Osmosis311: If you can confirm this works for you without any odd side-effects, I'll publish a new npm-package for it.

**Comment by Osmosis311** [comment]
@josteink it works, thank you very much!  Sorry for the delay!

**Comment by josteink** [comment]
Sorry about the delay here too.

Back from vacation now, and I've pushed a new package.

I'll consider this issue closed then :)
